pkg/placement/leadership.go

```diff
@@ -0,0 +1,150 @@
+package placement
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+)
+
+const barrierWriteTimeout = 2 * time.Minute
+
+// MonitorLeadership is used to monitor if we acquire or lose our role
+// as the leader in the Raft cluster. There is some work the leader is
+// expected to do, so we must react to changes
+//
+// reference: https://github.com/hashicorp/consul/blob/master/agent/consul/leader.go
+func (p *Service) MonitorLeadership(parentCtx context.Context) error {
+	if p.closed.Load() {
+		return errors.New("placement is closed")
+	}
+
+	ctx, cancel := context.WithCancel(parentCtx)
+	p.wg.Add(1)
+	go func() {
+		defer p.wg.Done()
+		defer cancel()
+		select {
+		case <-parentCtx.Done():
+		case <-p.closedCh:
+		}
+	}()
+
+	raft, err := p.raftNode.Raft(ctx)
+	if err != nil {
+		return err
+	}
+	var (
+		leaderCh                    = raft.LeaderCh()
+		oneLeaderLoopRun            sync.Once
+		loopNotRunning              = make(chan struct{})
+		leaderCtx, leaderLoopCancel = context.WithCancel(ctx)
+	)
+	close(loopNotRunning)
+
+	defer func() {
+		leaderLoopCancel()
+		<-loopNotRunning
+	}()
+
```

Comment:
We have three wrapped contexts here in this function, as well as double `select`s, `once`s, `defers` etc. which can be tricky to follow. Would using a `concurrency.RunnerManager` per `for` loop help?